internal/internal_workflow_client.go

```diff
@@ -139,6 +141,49 @@ func (wc *workflowClient) StartWorkflow(
 	return executionInfo, nil
 }
 
+// RunWorkflow starts a workflow execution and wait until this workflow reaches the end state, such as
+// workflow finished successfully or timeout.
+// The user can use this to start using a functor like below and get the workflow execution result, as encoded.Value
+// Either by
+//     RunWorkflow(options, "workflowTypeName", input)
+//     or
+//     RunWorkflow(options, workflowExecuteFn, arg1, arg2, arg3)
+// NOTE: the context.Context should have a fairly large timeout, since workflow execution may take a while to be finished
+func (wc *workflowClient) RunWorkflow(ctx context.Context, options StartWorkflowOptions, workflow interface{}, args ...interface{}) (encoded.Value, error) {
+
+	// start the workflow execution
+	executionInfo, err := wc.StartWorkflow(ctx, options, workflow, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	// wait until the workflow execution to finish by long poll on the last event
+	var history *s.History
+	var closeEvent *s.HistoryEvent
+	history, err = wc.PollWorkflowHistory(ctx, executionInfo.ID, executionInfo.RunID, s.HistoryEventFilterTypeCloseEvent)
+	if err != nil {
+		return nil, NewRunWorkflowError(executionInfo.RunID, "Error when tracking workflow history: "+err.Error())
+	}
+
+	closeEvent = history.Events[0]
+	errorMessage := ""
+	switch closeEvent.GetEventType() {
+	case s.EventTypeWorkflowExecutionCompleted:
+		return EncodedValue(closeEvent.WorkflowExecutionCompletedEventAttributes.Result), nil
+	case s.EventTypeWorkflowExecutionCanceled:
+		errorMessage = "Workflow execution is cancelled."
```

Comment:
It would be good if we could expose that error type as an enum field of the RunWorkflowError.